tools/windows/DatadogAgentInstaller/CustomActions/UserCustomActions.cs

```diff
@@ -471,10 +499,358 @@ private ActionResult AddUser()
+                _registryServices.OpenRegistryKey(Registries.LocalMachine, Constants.DatadogAgentRegistryKey);
+            var domain = subkey.GetValue("installedDomain")?.ToString();
+            var user = subkey.GetValue("installedUser")?.ToString();
+            if (string.IsNullOrEmpty(domain) || string.IsNullOrEmpty(user))
+            {
+                throw new Exception("Agent user information is not in registry");
+            }
+
+            var name = $"{domain}\\{user}";
+            _session.Log($"Found agent user information in registry {name}");
+            var userFound = _nativeMethods.LookupAccountName(name,
+                out _,
+                out _,
+                out var securityIdentifier,
+                out _);
+            if (!userFound || securityIdentifier == null)
+            {
+                throw new Exception($"Could not find account for user {name}.");
+            }
+
+            _session.Log($"Found previous agent user {name} ({securityIdentifier})");
+            return securityIdentifier;
+        }
+
+        /// <summary>
+        /// Recursively enables inheritance on files in APPLICATIONDATADIRECTORY
+        /// Removes any redundant explicit ACEs for the agent user
+        /// If changing the agent user, removes the previous agent user as owner of any files in APPLICATIONDATADIRECTORY
+        /// </summary>
+        private void EnablePermissionInheritance()
+        {
+            // Ensure that inheritance is enabled on all files/folders in the config directory.
+            // This is important so that @SetBaseInheritablePermissions are applied and the agent
+            // user has access where it needs it.
+            // If changing agent user, remove old user as owner/group from all files/folders
+            foreach (var filePath in Directory.EnumerateFileSystemEntries(
+                         _session.Property("APPLICATIONDATADIRECTORY"),
+                         "*.*", SearchOption.AllDirectories))
+            {
+                FileSystemSecurity fileSystemSecurity =
+                    _fileSystemServices.GetAccessControl(filePath, AccessControlSections.All);
+                bool changed = false;
+
+                // If changing agent user, remove old user as owner/group from all files/folders
+                if (previousDdAgentUserSID != null && previousDdAgentUserSID != ddAgentUserSID)
+                {
+                    changed |= RemoveOwnerGroup(filePath, fileSystemSecurity, previousDdAgentUserSID);
+                }
+
+                // Ensure that inheritance is enabled on all files/folders in the config directory
+                if (fileSystemSecurity.AreAccessRulesProtected)
+                {
+                    // enable inheritance
+                    fileSystemSecurity.SetAccessRuleProtection(false, true);
+                    changed = true;
+
+                    changed |= RemoveRedundantExplicitAccess(filePath, fileSystemSecurity);
```

Comment:
I was torn over this. on one hand the `changed = true` makes the `changed |=` redundant, but on the other hand I'm nervous about ignoring the return value of `RemoveRedundantExplicitAccess` just because of some code that's near it right now but could be changed later and the `changed |=` mistakenly not added back in.